python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,251 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Utility module for various conversions around PrimitiveType implementations
+
+This module enables:
+    - Converting partition strings to built-in python objects
+    - Converting a value to a byte buffer
+    - Converting a byte buffer to a value
+
+Note:
+    Conversion logic varies based on the PrimitiveType implementation. Therefore conversion functions
+    are defined here as generic functions using the @singledispatch decorator. For each PrimitiveType
+    implementation, a concrete function is registered for each generic conversion function. For PrimitiveType
+    implementations that share the same conversion logic, registrations can be stacked.
+
+Example (registering `from_partition_to_py` logic for BooleanType):
+    >>> @from_partition_value_to_py.register(BooleanType)
+    >>> def _(type_, partition_value_as_str: str):
+            if partition_value_as_str is None:
+                return False
+            return partition_value_as_str.lower() == "true"
+
+Example (stacking registrations):
+   >>> @from_partition_value_to_py.register(FloatType)
+   >>> @from_partition_value_to_py.register(DoubleType)
+   >>> def _(type_, partition_value_as_str: str):
+           return float(partition_value_as_str)
+"""
+import struct
+import uuid
+from decimal import ROUND_HALF_UP, Decimal
+from functools import singledispatch
+from typing import Any
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+HIVE_NULL = "__HIVE_DEFAULT_PARTITION__"
+
+
+@singledispatch
+def from_partition_value_to_py(type_, partition_value_as_str: str):
+    """A generic function function which converts a partition string to a python built-in
+
+    Args:
+        type_(PrimitiveType): An implementation of the PrimitiveType base class
+        partition_value_as_str(str): A string representation of a partition value
+    """
+    if partition_value_as_str is None or partition_value_as_str == HIVE_NULL:
+        return None
+    raise TypeError(f"Cannot convert partition string to python built-in, type {type_} not supported: '{partition_value_as_str}'")
+
+
+@from_partition_value_to_py.register(BooleanType)
+def _(type_, partition_value_as_str: str):
```

Comment:
hey @samredai , I appreciate your introduction of this dynamic overloading. I tried to adopt it in #4262 and looks like mypy is not happy about it as the method `_` now is being redefined multiple times with error code `no-redef`, do you know if we need to configure anything in mypy.ini or tox.ini to work around this issue?

Reply:
Basically this is what I saw

```
(venv) ➜  python git:(master) ✗ mypy --no-implicit-optional --show-error-codes  --config tox.ini src
tox.ini: No [mypy] section in config file
src/iceberg/expression/literals.py:27: error: Library stubs not installed for "pytz" (or incompatible with Python 3.8)  [import]
src/iceberg/expression/literals.py:27: note: Hint: "python3 -m pip install types-pytz"
src/iceberg/expression/literals.py:27: note: (or run "mypy --install-types" to install all missing stub packages)
src/iceberg/expression/literals.py:27: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
src/iceberg/expression/literals.py:109: error: Name "_of" already defined on line 104  [no-redef]
src/iceberg/expression/literals.py:119: error: Name "_of" already defined on line 104  [no-redef]
src/iceberg/expression/literals.py:129: error: Name "_of" already defined on line 104  [no-redef]
src/iceberg/expression/literals.py:134: error: Name "_of" already defined on line 104  [no-redef]
src/iceberg/expression/literals.py:139: error: Name "_of" already defined on line 104  [no-redef]
src/iceberg/expression/literals.py:144: error: Name "_of" already defined on line 104  [no-redef]
src/iceberg/expression/literals.py:149: error: Name "_of" already defined on line 104  [no-redef]
```

Reply:
It looks like you're using your locally installed version of mypy which I'm guessing is earlier than version `0.920` which is when singledispatch was added to the whitelist. The `tox` suite shouldn't run into this since it uses the latest version of mypy (currently `0.931`).

Also although `@singledispatch` was added in python 3.4, `@singledispatchmethod` wasn't added until python 3.8. Fortunately it was a very small addition and it was back-ported to python 3.7, so you'll just have to add `singledispatch` to the `deps` in the tox.ini file and the pyproject.toml file.